tests/commands/test_store_client.py

```diff
@@ -434,3 +438,116 @@ def test_client_errorparsing_bad_structure():
+
+        # check monitor is properly built
+        assert isinstance(monitor.encoder, MultipartEncoder)
+        srctype, fh, ctype = monitor.encoder.fields['binary']
+        assert srctype == 'filename'
+        assert fh.name == str(test_filepath)
+        assert ctype == "application/octet-stream"
+
+        content = json.dumps(dict(successful=True, upload_id='test-upload-id'))
+        return FakeResponse(content=content, status_code=200)
+
+    with patch('charmcraft.commands.store.client._storage_push', fake_pusher):
+        Client().push(test_filepath)
+
+    # check proper logs
+    expected = [
+        "Starting to push {}".format(str(test_filepath)),
+        "Uploading bytes ended, id test-upload-id",
+    ]
+    assert expected == [rec.message for rec in caplog.records]
+
+
+def test_client_push_response_not_ok(tmp_path):
+    """Didn't get a 200 from the Storage."""
+    # fake some bytes to push
+    test_filepath = tmp_path / 'supercharm.bin'
+    with test_filepath.open('wb') as fh:
+        fh.write(b"abcdefgh")
+
+    with patch('charmcraft.commands.store.client._storage_push') as mock:
+        mock.return_value = FakeResponse(content='had a problem', status_code=500)
+        with pytest.raises(CommandError) as cm:
+            Client().push(test_filepath)
+        assert str(cm.value) == "Failure when pushing file: [500] 'had a problem'"
+
+
+def test_client_push_response_unsuccessful(tmp_path):
+    """Didn't get a 200 from the Storage."""
+    # fake some bytes to push
+    test_filepath = tmp_path / 'supercharm.bin'
+    with test_filepath.open('wb') as fh:
+        fh.write(b"abcdefgh")
+
+    with patch('charmcraft.commands.store.client._storage_push') as mock:
+        raw_content = dict(successful=False, upload_id=None)
+        mock.return_value = FakeResponse(content=json.dumps(raw_content), status_code=200)
+        with pytest.raises(CommandError) as cm:
+            Client().push(test_filepath)
+        assert str(cm.value) == "Server error when pushing file: " + repr(raw_content)
+
+
+def test_storage_push_succesful():
+    """Bytes are properly pushed to the Storage."""
+    ctype = "application/octet-stream"
+    test_monitor = MultipartEncoderMonitor(MultipartEncoder(
+        fields={"binary": ("filename", 'somefile', ctype)}))
+
+    with patch('requests.Session') as mock:
+        _storage_push(test_monitor)
+
+    # check request was properly called
+    url = 'https://storage.staging.snapcraftcontent.com/unscanned-upload/'
+    headers = {
+        'Content-Type': ctype,
+        'Accept': 'application/json',
+        'User-Agent': build_user_agent(),
+    }
+    assert mock().request.called_once_with('POST', url, headers=headers, data=test_monitor)
+
+    # check the retries were properly setup
+    (protocol, adapter), _ = mock().mount.call_args
+    assert protocol == 'https://'
+    assert isinstance(adapter, HTTPAdapter)
+    assert adapter.max_retries.backoff_factor == 2
```

Comment:
Did you want a test that we actually do a retry?
Should we be logging a warning of some sort if we have to retry?
What happens with the progress bar if we get a retry after uploading 25% of the file?

Reply:
I don't want to test that "retry works", as I'm not implementing that (just using the lib).

There's no risk for the progress bar, as it uses "total bytes read" (it's not advancing the bar as "per ticks").